Allow setScreenshareActive to override WKWebView muting

youennf opened the pull request:
#### 2547c620af570749a7942d4ea8ecbbdbd4b259d4
<pre>
Allow setScreenshareActive to override WKWebView muting
<a href="https://rdar.apple.com/137410714">rdar://137410714</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=280960">https://bugs.webkit.org/show_bug.cgi?id=280960</a>

Reviewed by Eric Carlson.

When WKWebView mutes screenshare, setScreenshareActive was not allowed to unmute.
We are introducing a new decidePolicyForScreenCaptureUnmute APIUIClient callback to allow WKWebView to decide what to do.
The default implementation is to deny unmuting.
UIDelegate exposes a [_webView:decidePolicyForScreenCaptureUnmuteForOrigin:initiatedByFrame:decisionHandler:] delegate.
If no delegate is implemented, a Allow/Deny prompt is used, via alertForPermission for MediaPermissionReason::ScreenCapture.

Drive-by fix in MediaProducer since this can confuse which state is actually muted.

Covered by added API test.

* Source/WebCore/page/MediaProducer.h:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::decidePolicyForScreenCaptureUnmute):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::decidePolicyForScreenCaptureUnmute):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::validateCaptureStateUpdate):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/GetDisplayMediaWindowAndScreen.mm:
(-[DisplayGUMMessageHandler userContentController:didReceiveScriptMessage:]):
(TestWebKitAPI::TEST(WebKit2, ToggleScreenshare)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/media-session-capture.html:
* Tools/TestWebKitAPI/cocoa/UserMediaCaptureUIDelegate.mm:
(-[UserMediaCaptureUIDelegate _webView:decidePolicyForScreenCaptureUnmuteForOrigin:initiatedByFrame:decisionHandler:]):

Canonical link: <a href="https://commits.webkit.org/284883@main">https://commits.webkit.org/284883@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/171b393f6d762ff30f5468065063f15c8ba79b10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70828 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50238 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23597 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22028 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58036 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21849 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/74927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14525 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73894 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45666 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61063 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42321 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18496 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/20370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64255 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18858 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76640 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15058 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18058 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15102 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61126 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/63745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15678 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11816 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5463 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91858 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46039 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/810 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/91858 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47111 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48392 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46853 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11655,9 +11655,24 @@ void WebPageProxy::validateCaptureStateUpdate(WebCore::UserMediaRequestIdentifie
+                if (!page)
+                    return;
+                page->m_mutedCaptureKindsDesiredByWebApp.remove(WebCore::MediaProducerMediaCaptureKind::Display);
+                page->m_mutedCaptureKindsDesiredByWebApp.remove(WebCore::MediaProducerMediaCaptureKind::SystemAudio);
```
Do we always want to unmute System Audio capture when we unmute Display Capture? Does muting one always mute the other?

youennf replied:
I removed it for now.
I think we might want to treat them together but we should do this consistently, and in a separate patch.

eric-carlson reviewed (approved):
Not: I think `decidePolicyForScreenCaptureUnmuting` would be a slightly better name

youennf committed:
Allow setScreenshareActive to override WKWebView muting
rdar://137410714
https://bugs.webkit.org/show_bug.cgi?id=280960

Reviewed by Eric Carlson.

When WKWebView mutes screenshare, setScreenshareActive was not allowed to unmute.
We are introducing a new decidePolicyForScreenCaptureUnmute APIUIClient callback to allow WKWebView to decide what to do.
The default implementation is to deny unmuting.
UIDelegate exposes a [_webView:decidePolicyForScreenCaptureUnmuteForOrigin:initiatedByFrame:decisionHandler:] delegate.
If no delegate is implemented, a Allow/Deny prompt is used, via alertForPermission for MediaPermissionReason::ScreenCapture.

Drive-by fix in MediaProducer since this can confuse which state is actually muted.

Covered by added API test.

* Source/WebCore/page/MediaProducer.h:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::decidePolicyForScreenCaptureUnmute):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::decidePolicyForScreenCaptureUnmute):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::validateCaptureStateUpdate):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/GetDisplayMediaWindowAndScreen.mm:
(-[DisplayGUMMessageHandler userContentController:didReceiveScriptMessage:]):
(TestWebKitAPI::TEST(WebKit2, ToggleScreenshare)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/media-session-capture.html:
* Tools/TestWebKitAPI/cocoa/UserMediaCaptureUIDelegate.mm:
(-[UserMediaCaptureUIDelegate _webView:decidePolicyForScreenCaptureUnmuteForOrigin:initiatedByFrame:decisionHandler:]):

Canonical link: https://commits.webkit.org/284883@main
